Fix SSA being silently disabled by lack of OpenAPI config

alexzielenski opened the pull request:
#### What type of PR is this?

/kind bug
/kind cleanup

#### What this PR does / why we need it:

We've previously had the condition where, despite SSA being enabled, it would be silently disabled if the apiserver library user did not provide OpenAPI config and models for SSA to create a type converter. Since SSA is enabled by default and locked to default, this is surprising. In fact, despite the top-level kube-apiserver making proper use of SSA by providing these models, many unit/integration tests are not testing the SSA path since they in some way do not provide the OpenAPI config/definitions.

This PR:

1.) Makes it an error to not provide openapi definitions for each API resource installed during startup
2.) Fixes up tests to now provide proper OpenAPI schemas for their test servers (previously they would be testing with SSA effectively disabled)
3.) Cleans up conditional statements about SSA being enabled that are no longer required

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
NONE
```

 Related: https://github.com/kubernetes/kubernetes/pull/114439

/sig api-machinery
/assign @apelisse

alexzielenski committed:
Throw error if OpenAPIV3 config is not provided

it is required for Server-SIde-Apply to function correctly (SSA is based on OpenAPI schemas)

cici37 commented:
/triage accepted

deads2k commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:
```diff
@@ -883,7 +880,9 @@ func NewDefaultAPIGroupInfo(group string, scheme *runtime.Scheme, parameterCodec
 func (s *GenericAPIServer) getOpenAPIModels(apiPrefix string, apiGroupInfos ...*APIGroupInfo) (*spec.Swagger, error) {
 	if s.openAPIConfig == nil {
-		return nil, nil
+		// SSA is GA (enabled by default) and requires OpenAPI config to be set
```
I agree.  In my opinion, to be a kube-compatible API at this point, it must support SSA. Aggregated apiservers serving their APIs must participate to continue updating past this point.

deads2k commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:
```diff
@@ -667,6 +668,11 @@ func (s preparedGenericAPIServer) NonBlockingRun(stopCh <-chan struct{}, shutdow
 
 // installAPIResources is a private method for installing the REST storage backing each api groupversionresource
 func (s *GenericAPIServer) installAPIResources(apiPrefix string, apiGroupInfo *APIGroupInfo, openAPIModels *spec.Swagger) error {
+	typeConverter, err := fieldmanager.NewTypeConverter(openAPIModels, false)
```
We used to have an instance per groupversion.  I'm not familiar enough with the fieldmanager to know if that matters offhand.

cc @apelisse

apelisse replied:
We don't need multiple copies of these, they have the same input parameters and so are all going to be exact copies anyway. They are expensive to build and keep a lot of allocated memory in-use. They are considered immutable (except for some internal caching). And they all have the schema for all the internal types anyway. We're making their construction A LOT cheaper in #114439, but it's still much better to just do it once.

deads2k commented:
Adding the requirement for openapi in order to add an API makes sense at this point so that clients can have predictable interactions.

/approve
/assign @apelisse

apelisse commented:
/lgtm
/approve

apelisse commented:
What's blocking this by the way?

alexzielenski commented:
https://github.com/kubernetes/kubernetes/pull/115324 changes to use openAPIV3Config instead of openAPIConfig, so I did not want to make it an error to not have `openAPIConfig` right before changing the requirement to being `openAPIV3Config`

apelisse reviewed (commented):
As mentioned offline, there are many places where we check if typeconverter is set, maybe these should be updated too

alexzielenski committed:
supply deduced typeconverter to apiserver tests

alexzielenski committed:
add error for attempting to initialize field manager with nil typeconverter

alexzielenski committed:
add OpenAPIV3 config to tests and server options

alexzielenski committed:
include apiextensions types in apiextensions generated openapi

k8s_tag_files_matching looks for a slash after its argument, so the current value doesnt match anything

also update codegen

this is required for apiextensions-apiserver tests. After fixing apiextensions server tests to use type-aware SSA (instead of erroneously using untyped SSA) there were errors since none of the apiextensions types were actually used in the openapi given to tests.

alexzielenski committed:
fix scale not being given gvk in CRD openapi schemas

relates to  #116311

alexzielenski committed:
remove checks for fieldmanager from handlers

it should never be nil

alexzielenski committed:
fixup patcher test to include field manager

apelisse commented:
I'm excited to see that thread revived.

apelisse commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:
```diff
@@ -977,7 +969,13 @@ func (s *GenericAPIServer) getOpenAPIModels(apiPrefix string, apiGroupInfos ...*
 	}
-	return openAPISpec, nil
+
+	typeConverter, err := managedfields.NewTypeConverter(openAPISpec, false)
+	if err != nil {
+		return nil, err
+	}
+
+	return typeConverter, nil
```
Could be `return managedfields.NewTypeConverter(openAPISpec, false)`

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/patch.go:
```diff
@@ -344,6 +344,11 @@ func (p *jsonPatcher) applyPatchToCurrentObject(requestContext context.Context,
+		// Provide a more informative error for the crash that would
+		// happen on the next line
+		panic("PatchOptions required but not provided")
+	}
```
Have you seen that panic? Clearly it shouldn't happen. I'm guessing that the added options below?

apelisse reviewed (commented):
Minor nits, thanks!

apelisse commented:
/lgtm
/approve

RainbowMango commented:
Hey PR author and approvers, @alexzielenski @deads2k @apelisse
Do you think this patch should be backported to release branches?
I'm asking because [this issue hasn't been fixed](https://github.com/kubernetes/kubernetes/blob/6d31f4b31ba1294c4844ec94ea2f67c5f8bf42f4/staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go#L962) on release-1.27 and might block people from bumping their K8S dependencies.

apelisse commented:
@RainbowMango I don't think it's likely we will back-port that change, but I also don't understand what problem you have that would require this to be back-ported for? Can you elaborate?

RainbowMango commented:
@apelisse Thanks, but please ignore me now. We had some block issues(OpenAPI related) during [the update of Kubernetes dependencies to v1.27.3](https://github.com/karmada-io/karmada/pull/3730), but now it has been solved.
